### ISSUE TITLE
KL clipping: before or after averaging?

### ISSUE THREAD

**Issue by vmoens** [danijar/dreamer#56]
We're trying to reproduce the results from dreamer in [torchrl](https://github.com/facebookresearch/rl). We're struggling to make sense of the following apparent contradiction:

From the paper, we read "We do not scale the KL regularizers (β = 1) but clip them below 3 free nats as in PlaNet." However, in the [code](https://github.com/danijar/dreamer/blob/56d4d444dfd0582b0e79dab80aebbea74c0ce40d/dreamer.py#L175), the nats are clamped _after_ being averaged.
To me the paper sentence seems to point to a loss that would be an expectation over a function (the clip) and not a function over an expectation. Intuitively, the first seems to make more sense to me but I guess that if dreamer is trained using the second approach this may have a substantial impact on performance and choice of hyperparameters. Clipping the mean basically means that every time the whole average is below 3 nats the KL is totally ignored. In the first, only the gradients that pass through examples with a KL below 3 nats will be cut, hence the KL is more likely to be taken in consideration.

Do you have any insight to share about this?

**Comment by vmoens** [comment]
I see that this has been asked in #53.
I'm happy to close the issue for now. We'll run some experiments on this and post the results here if anything comes up.
